fcrepo4/fcrepo4
# Moves the search index into its own table.

Adds support for searching by modified and created dates.
Adds support for customizing the output fields of your search.
By default all fields returned.

Resolves:
https://jira.lyrasis.org/browse/FCREPO-3349
https://jira.lyrasis.org/browse/FCREPO-3353
https://jira.lyrasis.org/browse/FCREPO-3354


* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
* Moves the search index into its own table.
* Adds support for searching by modified and created dates.
* Adds support for customizing the output fields of your search.
* The index is kept up to date now via the internal event bus.
* Integration tests cover the following features:
   1. search by resources with modified or created date <, >, >=, or <= . Ensure that "before", "after" and "before and after" (ie between) are supported.
   2. compound queries are now supported: ie your can specify multiple conditions.

# How should this be tested?
Create a bunch of resources
```
for i in `seq 1 50`; do curl -X PUT -u fedoraAdmin:fedoraAdmin  "http://localhost:8080/rest/resource-$i"; done;
```
See that the modified and created dates are returned by default (`jq` makes the json pretty):
```
curl  -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search" | jq 
```
Return all available fields in results: (should be identical to the previous results )
```
curl -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search?fields=*" | jq
```
Return only the fedora_id and modified in results:
```
curl -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search?fields=fedora_id,modified" | jq
```
Return results matching fedora_id and created after using YYYY-mm-dd format. Should return one result
```
curl -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search?condition=fedora_id%3Dresource-1&condition=created%3E2020-06-18&fields=fedora_id,modified" | jq
```
Return results matching fedora_id and created before using YYYY-mm-dd format. Should return no  results.
```
curl -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search?condition=fedora_id%3Dresource-1&condition=created%3C2020-06-18&fields=fedora_id,modified" | jq
```
Return results matching fedora_id and modified after using ISO date format. Should return one result
```
curl -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:search?condition=fedora_id%3Dresource-1&condition=modified%3E2020-06-18T19:04:10.802Z&fields=fedora_id,modified" | jq
```

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo4/committers
